compiler/rustc_middle/src/ty/mod.rs

```diff
@@ -605,6 +605,14 @@ pub enum PredicateKind<'tcx> {
     ///
     /// Only used for Chalk.
     TypeWellFormedFromEnv(Ty<'tcx>),
+
+    /// Represents a hidden type assignment for an opaque type.
+    /// Such obligations get processed by checking whether the item currently being
+    /// type-checked may acually define it.
+    /// The arguments may be in any order, but one of them must be an opaque type.
+    /// The order and the boolean is used for resolving which opaque type defines
+    /// the other if both are opaque types and for diagnostics.
+    OpaqueType(Ty<'tcx>, Ty<'tcx>, bool),
```

Comment:
Why not just make the left be the opaque type being defined, and get rid of the bool?

Reply:
I think I tried that but got messed up diagnostics, but I'll try again